### PR TITLE
Ordering price information improvements.

### PR DESCRIPTION
**Pull request by FernandoOjeda**
Ordering price information improvements https://github.com/softlayer/softlayer-python/issues/1299.

Implemented:
  
    slcli hw create-options
    slcli order item-list

Things to take into implement:

    Hourly and Monthly prices
    Capacity restricted prices
    User can pick a location to display prices in

### CODE REVIEW

**Comment by FernandoOjeda** [comment]
I think the `slcli vs create-optoins` has to be improve the same that `slcli hw create-options` to add the prices information:

image

Or maybe I can add a new request to shows the item prices in another table and not to add then to the existing table.

**Comment by allmightyspiff** [comment]
Yeah I agree about `slcli vs|hw create-options`, I'll have to think about it a bit and decide how I want the output to look while I review your current changes.

File: SoftLayer/CLI/hardware/create_options.py [hunk 1]
```diff
@@ -4,13 +4,17 @@
 import click
 
 from SoftLayer.CLI import environment
+from SoftLayer.CLI import exceptions
 from SoftLayer.CLI import formatting
 from SoftLayer.managers import hardware
 
 
 @click.command()
+@click.option('--prices', '-p', default=False, help='Filter Item Prices, prices(DEFAULT False)')
```
Comment (allmightyspiff) [inline SoftLayer/CLI/hardware/create_options.py:13]: lets combine prices and location.
so if I do `--prices` I get all default price output. if I do `--prices mex01` I get mex01 prices for example. Might require using a callback function for `--prices` with no value after... not sure. 

Also, when using a location price, the default price should still show up if there isn't a location based one.

File: SoftLayer/CLI/hardware/create_options.py [hunk 2]
```diff
@@ -22,41 +26,171 @@ def cli(env):
...
+    tables.append(preset_prices_table)
+
+
+def _os_prices_table(operating_systems, tables):
+    """Shows Server Operating Systems prices cost and capacity restriction.
+
+    :param [] operating_systems: List of Hardware Server operating systems.
+    :param tables: Table formatting.
+    """
+    os_prices_table = formatting.Table(['OS Key', 'Hourly', 'Monthly', 'capacityRestrictionMaximum',
```
Comment (allmightyspiff) [inline SoftLayer/CLI/hardware/create_options.py:101]: To make this easier to read, lets make the capacity restriction part a bit more condensed for all tables with this field

```
:................................................................................:
: Operating Systems Prices                                                       :
:.............................................:........:.........:...............:
: OS Key                                      : Hourly : Monthly : Restriction   : 
:.............................................:........:.........:...............:
: OS_CENTOS_6_X_32_BIT                        : 0      : 0       : None          :
: OS_WINDOWS_2008_FULL_WEB_64_BIT_R2_SP1      : -      : 11      : 1-1 PROCESSOR :
: OS_WINDOWS_2008_FULL_WEB_64_BIT_R2_SP1      : -      : 22      : 2-2 PROCESSOR :
: OS_WINDOWS_2012_FULL_STD_64_BIT             : .056   : 36      : 2-4 CORE      :
: OS_WINDOWS_2012_FULL_STD_64_BIT             : .112   : 72      : 6-16 CORE     :
:.............................................:........:.........:...............:
```

File: SoftLayer/CLI/hardware/create_options.py [hunk 3]
```diff
@@ -22,41 +26,171 @@ def cli(env):
...
+    """Shows Server Preset options prices.
+
+    :param [] sizes: List of Hardware Server sizes.
+    :param tables: Table formatting.
+    """
+    preset_prices_table = formatting.Table(['Size', 'Value', 'Hourly', 'Monthly'], title="Sizes Prices")
+    preset_prices_table.sortby = 'Value'
+    preset_prices_table.align = 'l'
+    for size in sizes:
+        preset_prices_table.add_row([size['name'], size['key'], size['hourlyRecurringFee'], size['recurringFee']])
```
Comment (allmightyspiff) [inline SoftLayer/CLI/hardware/create_options.py:91]: recurringFees should have as many decimals in their output as from the API output. Which I think is 4 decimal places.

File: SoftLayer/CLI/order/item_list.py [hunk 4]
```diff
@@ -3,19 +3,25 @@
...
 COLUMNS = ['category', 'keyName', 'description', 'priceId']
+COLUMNS_ITEM_PRICES = ['keyName', 'priceId', 'Hourly', 'Monthly', 'CRMax', 'CRMim', 'CRType']
+COLUMNS_ITEM_PRICES_LOCATION = ['keyName', 'priceId', 'Hourly', 'Monthly', 'CRMax', 'CRMim', 'CRType']
 
 
 @click.command()
 @click.argument('package_keyname')
 @click.option('--keyword', help="A word (or string) used to filter item names.")
 @click.option('--category', help="Category code to filter items by")
+@click.option('--prices', '-p', default=False, help='Filter Item Prices, prices(DEFAULT False)')
```
Comment (allmightyspiff) [inline SoftLayer/CLI/order/item_list.py:20]: Same comment as `sl hw create-options` about combining these.

File: SoftLayer/CLI/order/item_list.py [hunk 5]
```diff
@@ -47,9 +55,22 @@ def cli(env, package_keyname, keyword, category):
     sorted_items = sort_items(items)
 
     categories = sorted_items.keys()
-    for catname in sorted(categories):
-        for item in sorted_items[catname]:
-            table.add_row([catname, item['keyName'], item['description'], get_price(item)])
+    if prices:
+        table = formatting.Table(COLUMNS_ITEM_PRICES, title="CRMax = CapacityRestrictionMaximum, "
```
Comment (allmightyspiff) [inline SoftLayer/CLI/order/item_list.py:59]: As with `create-options` combine the capacity columns in to a single column called `Restrictions` with the format `{Min}-{Max} {TYPE}`

**Comment by allmightyspiff** [comment]
Thanks, this is a big improvement I think. We'll tackle the `vs create-options` in another issue.

Big things to change:

1. `--prices` should take either 0 arguments, meaning default prices. Or 1 argument, meaning location specific prices
2. Capacity restrictions should be combined into a single column to make it easier to read
3. When using location specific pricing, default pricing should still be shown if there is no location pricing. Hopefully that isn't too complicated to add.

File: SoftLayer/CLI/hardware/create_options.py (hunk 1 above)
Comment (FernandoOjeda) [reply SoftLayer/CLI/hardware/create_options.py]: Done.

File: SoftLayer/CLI/hardware/create_options.py (hunk 2 above)
Comment (FernandoOjeda) [reply SoftLayer/CLI/hardware/create_options.py]: Done.

File: SoftLayer/CLI/hardware/create_options.py (hunk 3 above)
Comment (FernandoOjeda) [reply SoftLayer/CLI/hardware/create_options.py:87]: recurringFees seems to be all integer values, I did not see any recurringFees with decimal values by api output and printing this option by the code.
Please could you check for me this.

File: SoftLayer/CLI/order/item_list.py (hunk 4 above)
Comment (FernandoOjeda) [reply SoftLayer/CLI/order/item_list.py]: Done.

File: SoftLayer/CLI/order/item_list.py (hunk 5 above)
Comment (FernandoOjeda) [reply SoftLayer/CLI/order/item_list.py]: Done.

File: SoftLayer/CLI/hardware/create_options.py (hunk 3 above)
Comment (FernandoOjeda) [reply SoftLayer/CLI/hardware/create_options.py:87]: Do you mean this?
image
image

Comment (FernandoOjeda) [reply SoftLayer/CLI/hardware/create_options.py]: Done.

**Comment by allmightyspiff** [comment]
## TESTS
✔️ `hw create-options --prices`
✔️ `hw create-options --prices dal13`
✔️ `hw create-options --prices tok02`
✔️ `order item-list --prices AMSTERDAM02  SINGLE_E31270_V6`
❌ `order item-list SINGLE_E31270_V6  --prices dal13`
I get the error `Package dal13 does not exist`, not quite sure from looking at the code why click isn't processing that field, maybe because you have `type=click.UNPROCESSED`?



## Future Issues
Once this is merged I'll make issues for these other improvements. This change set is already getting kinda large and complex.

- `hw create-options --prices tok02` lists a few 0$ servers. I think that is because those are not available in tok02, as the others are priced correctly. These should be removed, or labeled as `not available` some way. We can solve that in a different issue though, since this one is already a big change.


```
$ ./slcli --format=table hardware create-options --prices tok02
:....................:
:    Datacenters     :
:............:.......:
: Datacenter : Value :
:............:.......:
: Tokyo 2    : tok02 :
:............:.......:
:.....................................................................................:
:                                     Sizes Prices                                    :
:...............................:.................................:........:..........:
: Size                          : Value                           : Hourly : Monthly  :
:...............................:.................................:........:..........:
: 1U 1270v3 32GB 2x2TB No-RAID  : 1U_1270_V3_2X2TB_NORAID         : 0.0000 : 0.0000   :
: 1U 1270v6 32GB 2x2TB No-RAID  : 1U_1270_V6_2X2TB_NORAID         : 0.0000 : 0.0000   :
: 1U 2174s x 64GB, 2x4TB, RAID1 : 1U_2174S_64GB_2X4TB_RAID_1      : 0.0770 : 57.0000  :
: 1U 4210s, 384GB, 2x4TB, RAID1 : 1U_4210S_384GB_2X4TB_RAID_1     : 0.0600 : 39.5500  :
: 1U 6248 384GB 1x3.8 SED JBOD  : 1U_6248S_384GB_1X3_8_SED_JBOD   : 0.4260 : 283.0000 :
: 1U 6248s, 384GB, 2x4TB, RAID1 : 1U_6248S_384GB_2X4TB_RAID_1     : 0.0770 : 57.0000  :
: 1U 6248 384 4x3.8 SED RAID-10 : 1U_6248S_384GB_4X_3_8TB_RAID_10 : 0.7400 : 492.0000 :
:...............................:.................................:........:..........:
```

- break out `order item-list`  by category, like `hw create-options`. Right now they are sorted by category, but I really like how `hw create-options` looks now. And Right Align the keyname/description fields to be easier to read.
- I'd like `item-list` to behave like `create-options` with the datacenter name as well, instead of the keyname. I think most customers expect `dal13` form, not the longer keyname form.
- I'd like for the location specific prices to just be in the first table (replacing their generic item when applicable).

**Comment by FernandoOjeda** [comment]
It is a good idea to open another issue for hw create-options --prices tok02 to avoid to lists the 0$ data.
And it would be good to open another issue to refactor the order item-list the same than hardware create-options.

**Pull request merged**